### ISSUE TITLE
py::cast<T> where T is a subclass of py::object is unchecked

### ISSUE THREAD

**Issue by ezyang** [pybind/pybind11#1108]
## Issue description

I expect something like `py::cast<py::bytes>(x)` to check if the `py::object x` is actually bytes. It does not.

## Reproducible example code

```
#include <pybind11/pybind11.h>
#include <iostream>
namespace py = pybind11;
int main() {
  Py_Initialize();
  auto gil = PyGILState_Ensure();
  PyObject *l = PyFloat_FromDouble(3.14);
  //PyObject *l = PyBytes_FromString("foo");
  auto n = py::cast<py::bytes>(l);
  std::cerr << "No problem yet...\n";
  std::cerr << static_cast<std::string>(n);
  PyGILState_Release(gil);
  return 0;
}
```

**Comment by bstaletic** [comment]
For the record, Boost.Python behaves the same. Boost expects you to run `n.check()` before using `n`. Pybind instead provides `py::isinstance<py::bytes>(n)`.

**Comment by YannickJadoul** [comment]
It can be considered as unexpected though. I'd vote for thinking this over, at least. I still get lost on when to use the constructor of such a subclass, or `py::cast`, or `py::isinstance` and a conversion method, or ...

**Comment by bstaletic** [comment]
This has been fixed in 2.6.0.

**Comment by YannickJadoul** [comment]
Should be #2349 that fixed this.
